python/ray/train/jax/jax_trainer.py

```diff
@@ -0,0 +1,205 @@
+from typing import Callable, Optional, Dict, Union, TYPE_CHECKING
+import logging
+
+from ray.air._internal.config import ensure_only_allowed_dataclass_keys_updated
+from ray.train.jax.config import JaxConfig
+from ray.train.trainer import GenDataset
+from ray.train.data_parallel_trainer import DataParallelTrainer
+from ray.air.config import ScalingConfig, RunConfig, DatasetConfig
+from ray.air.checkpoint import Checkpoint
+from ray.util import PublicAPI
+
+if TYPE_CHECKING:
+    from ray.data.preprocessor import Preprocessor
+
+logger = logging.getLogger(__name__)
+
+
+@PublicAPI(stability="alpha")
+class JaxTrainer(DataParallelTrainer):
+    """A Trainer for data parallel jax training.
+
+    This Trainer runs the function ``train_loop_per_worker`` on multiple Ray
+    Actors. These actors already have the necessary jax process group already
+    configured for distributed jax training.
+
+    The ``train_loop_per_worker`` function is expected to take in either 0 or 1
+    arguments:
+
+    .. code-block:: python
+
+        def train_loop_per_worker():
+            ...
+
+    .. code-block:: python
+
+        def train_loop_per_worker(config: Dict):
+            ...
+
+    If ``train_loop_per_worker`` accepts an argument, then
+    ``train_loop_config`` will be passed in as the argument. This is useful if you
+    want to tune the values in ``train_loop_config`` as hyperparameters.
+
+    If the ``datasets`` dict contains a training dataset (denoted by
+    the "train" key), then it will be split into multiple dataset
+    shards that can then be accessed by ``session.get_dataset_shard("train")`` inside
+    ``train_loop_per_worker``. All the other datasets will not be split and
+    ``session.get_dataset_shard(...)`` will return the the entire Dataset.
+
+    Note: Here are some details about how to use the ``JaxTrainer`` properly.
+
+    * cluster scaling config:
+        * For the GPU distributed cases, the inner-device communication is handled
+            internally by Jax; the inter-device communication is set up by the
+            ``JaxTrainer``. Therefore, the ``num_workers`` is set to be the number of
+            nodes; and ``num_gpus_per_worker`` is set to be the number of gpus on each
+            node. For the current experimental version, we only support the homogeneous
+            distributed case, i.e. all the nodes have the same number of gpus. This is
+            also the optimal / balanced case in terms of performance.
+        * For the TPU distributed cases, the multi-device distributed training only
+            supports the TPU pods for a) the TPU pods can host up to several thousands
+            of TPU cores; b) the communication speed is also optimal for the TPU pods.
+            Therefore, the ``num_workers`` is set to be the number of TPU-VMs
+            (say ``num_workers=4`` for TPU-pod v2-32); and
```

Comment:
does ray cluster launcher support tpu pods @Yard1?

currently we can only spin up multiple TPU devices, but these do not have high speed inter-connect is that right?

Reply:
That is correct. @simon-mo I believe worked on TPU pod support but I am not sure how far along that is

Reply:
Got it, can we update this doc string then?